File: src/skid_steer_localization/src/kalman_filter.cpp
#include "skid_steer_localization/kalman_filter.hpp"

KalmanFilter::KalmanFilter(const std::string &name): Node(name), mean{0.0}, variance{1000}, imu_vyaw{0.0},
                                                     last_vyaw{0.0}, motion{0.0}, is_first_odom{true},
                                                     motion_variance{4.0}, measurement_variance{0.5}
{
    odom_subscriber = create_subscription<nav_msgs::msg::Odometry>("skid_steer_controller/odom_noisy", 10, 
                      std::bind(&KalmanFilter::odom_callback, this, std::placeholders::_1));
    imu_subscriber = create_subscription<sensor_msgs::msg::Imu>("imu/out", 100, 
                      std::bind(&KalmanFilter::imu_callback, this, std::placeholders::_1));
    odom_publisher = create_publisher<nav_msgs::msg::Odometry>("skid_steer_controller/odom_filtered", 10);                                   
}

void KalmanFilter::imu_callback(const sensor_msgs::msg::Imu &imu){
    imu_vyaw = imu.angular_velocity.z;

}

void KalmanFilter::odom_callback(const nav_msgs::msg::Odometry &odom){
    odom_filtered = odom;
    if(is_first_odom){
        mean = odom.twist.twist.angular.z;
        is_first_odom = false;
        return;
    }

    motion = odom.twist.twist.angular.z - last_vyaw;


    state_prediction();
    update_measurements();
    odom_filtered.twist.twist.angular.z = mean;
    odom_publisher->publish(odom_filtered);
    
}

void KalmanFilter::update_measurements(){
    mean = (measurement_variance * mean + variance * imu_vyaw) / (variance + measurement_variance);
    variance = (variance * measurement_variance) / (variance + measurement_variance);
}

void KalmanFilter::state_prediction(){
    mean = mean + motion;
    variance = variance + motion_variance;
}


int main(int argc, char *argv[]){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<KalmanFilter>("kalman_filter");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: install/skid_steer_localization/local/lib/python3.10/dist-packages/skid_steer_localization/kalman_filter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
from sensor_msgs.msg import Imu
class KalmanFilter(Node):
    def __init__(self):
        super().__init__("kalman_filter")

        self.odom_subscriber = self.create_subscription(Odometry, "skid_steer_controller/odom_noisy", self.odom_callback, 10)
        self.imu_subscriber = self.create_subscription(Imu, "imu/out", self.imu_callback, 10)
        self.odom_publisher = self.create_publisher(Odometry, "skid_steer_controller/odom_filtered", 10)

        self.mean = 0.0
        self.variance = 1000.0
        self.imu_vyaw = 0.0
        self.is_first_odom = True
        self.last_vyaw = 0.0

        self.motion = 0.0
        self.odom_filtered = Odometry()
        
        self.motion_variance = 4.0
        self.measurement_variance = 0.5


    def update_measurements(self):
        self.mean = (self.measurement_variance * self.mean * self.variance * self.imu_vyaw) / (self.variance + self.measurement_variance)
        self.variance = (self.variance * self. measurement_variance) / (self.variance + self.measurement_variance)


    def state_prediction(self):
        self.mean = self.mean + self.motion
        self.variance = self.variance + self.motion_variance

    def imu_callback(self, imu):    
        self.imu_vyaw = imu.angular_velocity.z


    def odom_callback(self, odom):
        self.odom_filtered = odom
        if self.is_first_odom:
            self.mean = odom.twist.twist.angular.z
            self.last_vyaw = odom.twist.twist.angular.z
            self.is_first_odom = False
            return
        self.motion = odom.twist.twist.angular.z - self.last_vyaw

        self.state_prediction()

        self.update_measurements()

        self.odom_filtered.twist.twist.angular.z = self.mean
        self.odom_publisher.publish(self.odom_filtered)


def main():
    rclpy.init()
    kalman_filter = KalmanFilter()
    rclpy.spin(kalman_filter)
    kalman_filter.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()